drivers/kw41zrf/kw41zrf_getset.c

```diff
@@ -108,35 +110,41 @@ void kw41zrf_set_pan(kw41zrf_t *dev, uint16_t pan)
     DEBUG("[kw41zrf] set pan to: 0x%x\n", pan);
 }
 
-void kw41zrf_set_addr_short(kw41zrf_t *dev, network_uint16_t addr)
+void kw41zrf_set_addr_short(kw41zrf_t *dev, const network_uint16_t *addr)
 {
     (void) dev;
     ZLL->MACSHORTADDRS0 = (ZLL->MACSHORTADDRS0 & ~ZLL_MACSHORTADDRS0_MACSHORTADDRS0_MASK) |
-        ZLL_MACSHORTADDRS0_MACSHORTADDRS0(addr.u16);
+        ZLL_MACSHORTADDRS0_MACSHORTADDRS0(addr->u16);
 }
 
-void kw41zrf_set_addr_long(kw41zrf_t *dev, eui64_t addr)
+void kw41zrf_set_addr_long(kw41zrf_t *dev, const eui64_t *addr)
 {
     (void) dev;
-    ZLL->MACLONGADDRS0_LSB = (uint32_t)addr.uint64.u64;
-    ZLL->MACLONGADDRS0_MSB = (addr.uint64.u64 >> 32);
+
+    /* network order */
+    eui64_t a;
+    memcpy(&a, addr, sizeof(a));
+    a.uint64.u64 = byteorder_swapll(a.uint64.u64);
```

Comment:
`eui64_t` is already in network byte order.

Reply:
I don't know how to reduce casting and swapping further without functional changes.

Reply:
I thought that with the new `luid_get_eui64()` function you should be able to drop the `byteorder_swapll()` entirely if your device expects network byte order.

Does
```C
    ZLL->MACLONGADDRS0_LSB = addr->uint32[0].u32;
    ZLL->MACLONGADDRS0_MSB = addr->uint32[1].u32;
```
not work?

Reply:
Without that swap unicast packet reception stops working so `ping6 ff02::1%7` gets no replies. The pings come out ok but the rx hardware is filtering the pongs because the long address doesn't match.

Reply:
I tried 

```patch
--- a/drivers/kw41zrf/kw41zrf.c
+++ b/drivers/kw41zrf/kw41zrf.c
@@ -41,11 +41,11 @@ static void kw41zrf_set_address(kw41zrf_t *dev)
     DEBUG("[kw41zrf] Set MAC address\n");
     eui64_t addr_long;
     network_uint16_t addr_short;
+
     /* get unique IDs to use as hardware addresses */
     luid_get_eui64(&addr_long);
     luid_get_short(&addr_short);
-    /* this places the unique bits in less significant places */
-    addr_long.uint64.u64 = byteorder_swapll(addr_long.uint64.u64);
+
     /* set short and long address */
     kw41zrf_set_addr_long(dev, &addr_long);
     kw41zrf_set_addr_short(dev, &addr_short);
diff --git a/drivers/kw41zrf/kw41zrf_getset.c b/drivers/kw41zrf/kw41zrf_getset.c
index 77070d7ad0..ef66e68251 100644
--- a/drivers/kw41zrf/kw41zrf_getset.c
+++ b/drivers/kw41zrf/kw41zrf_getset.c
@@ -121,13 +121,8 @@ void kw41zrf_set_addr_long(kw41zrf_t *dev, const eui64_t *addr)
 {
     (void) dev;
 
-    /* network order */
-    eui64_t a;
-    memcpy(&a, addr, sizeof(a));
-    a.uint64.u64 = byteorder_swapll(a.uint64.u64);
-
-    ZLL->MACLONGADDRS0_LSB = a.uint32[0].u32;
-    ZLL->MACLONGADDRS0_MSB = a.uint32[1].u32;
+    ZLL->MACLONGADDRS0_LSB = addr->uint32[0].u32;
+    ZLL->MACLONGADDRS0_MSB = addr->uint32[1].u32;
 }
```

and ping is still working.

Reply:
Ok in that case I get the same result but then if I use ifconfig to set the long address it sets the address backwards.

Reply:
But then you have to swap it in the `case NETOPT_ADDRESS_LONG:` cases :weary: 
IMHO that's still a bit less confusing than swapping it back and forth, but it makes me wonder if `eui64_t` is wrong?

So how about
<details>
<summary>this patch</summary>

```patch
diff --git a/drivers/kw41zrf/kw41zrf.c b/drivers/kw41zrf/kw41zrf.c
index d5ac75fa6b..dcf4b9dc47 100644
--- a/drivers/kw41zrf/kw41zrf.c
+++ b/drivers/kw41zrf/kw41zrf.c
@@ -41,11 +41,11 @@ static void kw41zrf_set_address(kw41zrf_t *dev)
     DEBUG("[kw41zrf] Set MAC address\n");
     eui64_t addr_long;
     network_uint16_t addr_short;
+
     /* get unique IDs to use as hardware addresses */
     luid_get_eui64(&addr_long);
     luid_get_short(&addr_short);
-    /* this places the unique bits in less significant places */
-    addr_long.uint64.u64 = byteorder_swapll(addr_long.uint64.u64);
+
     /* set short and long address */
     kw41zrf_set_addr_long(dev, &addr_long);
     kw41zrf_set_addr_short(dev, &addr_short);
diff --git a/drivers/kw41zrf/kw41zrf_getset.c b/drivers/kw41zrf/kw41zrf_getset.c
index 77070d7ad0..285586eceb 100644
--- a/drivers/kw41zrf/kw41zrf_getset.c
+++ b/drivers/kw41zrf/kw41zrf_getset.c
@@ -121,13 +121,8 @@ void kw41zrf_set_addr_long(kw41zrf_t *dev, const eui64_t *addr)
 {
     (void) dev;
 
-    /* network order */
-    eui64_t a;
-    memcpy(&a, addr, sizeof(a));
-    a.uint64.u64 = byteorder_swapll(a.uint64.u64);
-
-    ZLL->MACLONGADDRS0_LSB = a.uint32[0].u32;
-    ZLL->MACLONGADDRS0_MSB = a.uint32[1].u32;
+    ZLL->MACLONGADDRS0_LSB = addr->uint32[0].u32;
+    ZLL->MACLONGADDRS0_MSB = addr->uint32[1].u32;
 }
 
 void kw41zrf_get_addr_short(kw41zrf_t *dev, network_uint16_t *addr)
@@ -142,9 +137,6 @@ void kw41zrf_get_addr_long(kw41zrf_t *dev, eui64_t *addr)
     (void) dev;
     addr->uint32[0] = (network_uint32_t)ZLL->MACLONGADDRS0_LSB;
     addr->uint32[1] = (network_uint32_t)ZLL->MACLONGADDRS0_MSB;
-
-    /* network order */
-    addr->uint64.u64 = byteorder_swapll(addr->uint64.u64);
 }
 
 int8_t kw41zrf_get_cca_threshold(kw41zrf_t *dev)
diff --git a/drivers/kw41zrf/kw41zrf_netdev.c b/drivers/kw41zrf/kw41zrf_netdev.c
index 70778f6536..369ed78580 100644
--- a/drivers/kw41zrf/kw41zrf_netdev.c
+++ b/drivers/kw41zrf/kw41zrf_netdev.c
@@ -566,6 +566,7 @@ int kw41zrf_netdev_get(netdev_t *netdev, netopt_t opt, void *value, size_t len)
         case NETOPT_ADDRESS_LONG:
             assert(len >= sizeof(eui64_t));
             kw41zrf_get_addr_long(dev, value);
+            *(uint64_t*)value = byteorder_swapll(*(uint64_t*)value);
             res = sizeof(eui64_t);
             break;
 
@@ -773,11 +774,14 @@ static int kw41zrf_netdev_set(netdev_t *netdev, netopt_t opt, const void *value,
             res = sizeof(const network_uint16_t);
             break;
 
-        case NETOPT_ADDRESS_LONG:
+        case NETOPT_ADDRESS_LONG: {
+            eui64_t addr;
             assert(len <= sizeof(const eui64_t));
-            kw41zrf_set_addr_long(dev, value);
+            addr.uint64.u64 = byteorder_swapll(*(uint64_t*)value);
+            kw41zrf_set_addr_long(dev, &addr);
             res = sizeof(const eui64_t);
             break;
+        }
 
         case NETOPT_NID:
             assert(len <= sizeof(const uint16_t));
```
</details>

Reply:
I had imagined that was the one place a swap shouldn't be necessary and based everything on that.

Anyway I'm entirely indifferent now about which way it goes so I'm content to patch it that way if you think it's best.

Reply:
I think it *feels* more right when the low lever functions just provide register access and the netdev adaption layer does whatever is necessary to accommodate the upper layer.